RISC-V: Enable IPI CPU Backtrace

uestc-gr opened the pull request:
当前riscv版本无法像arm和x86那样在oops故障时打印所有cpu堆栈，故障发生时只能打印当前cpu堆栈回溯。
这个补丁在oops或lookup等故障排查时相当重要，在sysctl中配置了kernel.oops_all_cpu_backtrace=1，这个补丁可以让故障发生时打印出所有cpu的堆栈回溯。

类似于

[ 229.185877] Sending NMI from CPU 5 to CPUs 0-4,6-7:
[ 229.187808] NMI backtrace for cpu 7
[ 229.188979] CPU: 7 PID: 0 Comm: swapper/7 Not tainted 6.6.25 https://github.com/RVCK-Project/rvck-olk/pull/20
[ 229.190026] Hardware name: riscv-virtio,qemu (DT)
[ 229.190947] epc : arch_local_irq_enable+0xc/0x14
[ 229.191988] ra : default_idle_call+0x3a/0x42
[ 229.193013] epc : ffffffff8007dbb0 ra : ffffffff80ed8120 sp : ff2000000013bed0
[ 229.194166] gp : ffffffff8202d8a0 tp : ff60000080289900 t0 : ff200000006d35c0
[ 229.195611] t1 : ffebffff4e9ac459 t2 : ffffffff818264a8 s0 : ff2000000013bee0
[ 229.196927] s1 : ffffffff8203a1c0 a0 : 0000000000000000 a1 : 0000000000000004
[ 229.198058] a2 : 0000000000000000 a3 : 0000000000001d6c a4 : ff600002f3b34000
[ 229.199242] a5 : 4000000000000000 a6 : ffebffff4e9ac45a a7 : ff60000274d622cb
[ 229.200451] s2 : 0000000000000007 s3 : 000000000000003f s4 : 0000000000000007
[ 229.201556] s5 : ff2000000013bf40 s6 : 1fe40000000277e0 s7 : 0000000000000000
[ 229.202680] s8 : 0000000000000000 s9 : ff60000080289900 s10: 0000000000000000
[ 229.203931] s11: 0000000000000000 t3 : ffebffff4e9ac459 t4 : ffebffff4e9ac45a
[ 229.204918] t5 : ffffffff818139a0 t6 : ffffffff818139e8
[ 229.205696] status: 0000000200000120 badaddr: 0000000000000000 cause: 8000000000000009
[ 229.206833] [] arch_local_irq_enable+0xc/0x14
[ 229.208202] [] default_idle_call+0x3a/0x42
[ 229.209377] [] do_idle+0x2c4/0x322
[ 229.210422] [] cpu_startup_entry+0x36/0x38
[ 229.211533] [] smp_callin+0xa4/0xd4


fixed: <https://github.com/RVCK-Project/rvck/issues/52>

oervci commented:
开始测试

oervci commented:
check patch done. log: https://jenkins.oerv.ac.cn/job/rvck-pipeline/job/check-patch/26/consoleFull

oervci commented:
Kernel build failed! log:https://jenkins.oerv.ac.cn/job/rvck-pipeline/job/rvck-kernel-build/47/consoleFull

oervci commented:
kunit test done. log:https://jenkins.oerv.ac.cn/job/rvck-pipeline/job/kunit-test/29/consoleFull

oervci commented:
开始测试

oervci commented:
check patch done. log: https://jenkins.oerv.ac.cn/job/rvck-pipeline/job/check-patch/28/consoleFull

oervci commented:
Kernel build success!

oervci commented:
kunit test done. log:https://jenkins.oerv.ac.cn/job/rvck-pipeline/job/kunit-test/31/consoleFull

oervci commented:
Lava check done! result url: https://lava.oerv.ac.cn/scheduler/job/192

uestc-gr committed:
RISC-V: Enable IPI CPU Backtrace

mainline inclusion
from mainline-v6.11-rc2
commit f15c21a3de1b5321ba4ae2c6c8f2e63b839d49fa
category: feature
bugzilla: #63

--------------------------------

Add arch_trigger_cpumask_backtrace() which is a generic infrastructure
for sampling other CPUs' backtrace using IPI.

The feature is used when lockups are detected or in case of oops/panic
if parameters are set accordingly.

Below is the case of oops with the oops_all_cpu_backtrace enabled.

$ sysctl kernel.oops_all_cpu_backtrace=1

triggering oops shows:
[  212.214237] NMI backtrace for cpu 1
[  212.214390] CPU: 1 PID: 610 Comm: in:imklog Tainted: G           OE      6.10.0-rc6 #1
[  212.214570] Hardware name: riscv-virtio,qemu (DT)
[  212.214690] epc : fallback_scalar_usercopy+0x8/0xdc
[  212.214809]  ra : _copy_to_user+0x20/0x40
[  212.214913] epc : ffffffff80c3a930 ra : ffffffff8059ba7e sp : ff20000000eabb50
[  212.215061]  gp : ffffffff82066f90 tp : ff6000008e958000 t0 : 3463303866660000
[  212.215210]  t1 : 000000000000005b t2 : 3463303866666666 s0 : ff20000000eabb60
[  212.215358]  s1 : 0000000000000386 a0 : 00007ff6e81df926 a1 : ff600000824df800
[  212.215505]  a2 : 000000000000003f a3 : 7fffffffffffffc0 a4 : 0000000000000000
[  212.215651]  a5 : 000000000000003f a6 : 0000000000000000 a7 : 0000000000000000
[  212.215857]  s2 : ff600000824df800 s3 : ffffffff82066cc0 s4 : 0000000000001c1a
[  212.216074]  s5 : ffffffff8206a5a8 s6 : 00007ff6e81df926 s7 : ffffffff8206a5a0
[  212.216278]  s8 : ff600000824df800 s9 : ffffffff81e25de0 s10: 000000000000003f
[  212.216471]  s11: ffffffff8206a59d t3 : ff600000824df812 t4 : ff600000824df812
[  212.216651]  t5 : ff600000824df818 t6 : 0000000000040000
[  212.216796] status: 0000000000040120 badaddr: 0000000000000000 cause: 8000000000000001
[  212.217035] [<ffffffff80c3a930>] fallback_scalar_usercopy+0x8/0xdc
[  212.217207] [<ffffffff80095f56>] syslog_print+0x1f4/0x2b2
[  212.217362] [<ffffffff80096e5c>] do_syslog.part.0+0x94/0x2d8
[  212.217502] [<ffffffff800979e8>] do_syslog+0x66/0x88
[  212.217636] [<ffffffff803a5dda>] kmsg_read+0x44/0x5c
[  212.217764] [<ffffffff80392dbe>] proc_reg_read+0x7a/0xa8
[  212.217952] [<ffffffff802ff726>] vfs_read+0xb0/0x24e
[  212.218090] [<ffffffff803001ba>] ksys_read+0x64/0xe4
[  212.218264] [<ffffffff8030025a>] __riscv_sys_read+0x20/0x2c
[  212.218453] [<ffffffff80c4af9a>] do_trap_ecall_u+0x60/0x1d4
[  212.218664] [<ffffffff80c56998>] ret_from_exception+0x0/0x64

Signed-off-by: Ryo Takakura <takakura@valinux.co.jp>
Link: https://lore.kernel.org/r/20240718093659.158912-1-takakura@valinux.co.jp
Signed-off-by: Palmer Dabbelt <palmer@rivosinc.com>
[Since Euler has modified the return type of arch_trigger_cpumask_backtrace in a
previous commit, we also need to make the same change.
Refer to “nmi: backtrace: Allow runtime arch specific override”.]
Signed-off-by: Gao Rui <gao.rui@zte.com.cn>

oervci commented:
开始测试

oervci commented:
check patch done. log: https://jenkins.oerv.ac.cn/job/rvck-pipeline/job/check-patch/35/consoleFull

oervci commented:
开始测试

oervci commented:
check patch done. log: https://jenkins.oerv.ac.cn/job/rvck-pipeline/job/check-patch/38/consoleFull

oervci commented:
kunit test done. log:https://jenkins.oerv.ac.cn/job/rvck-pipeline/job/kunit-test/38/consoleFull

oervci commented:
kunit test done. log:https://jenkins.oerv.ac.cn/job/rvck-pipeline/job/kunit-test/39/consoleFull

wangliu-iscas commented:
/check

oervci commented:
开始测试

oervci commented:
Kernel build failed! log:https://jenkins.oerv.ac.cn/job/rvck-pipeline/job/rvck-kernel-build/84/consoleFull

oervci commented:
kunit test done. log:https://jenkins.oerv.ac.cn/job/rvck-pipeline/job/kunit-test/61/consoleFull

oervci commented:
check patch done. log: https://jenkins.oerv.ac.cn/job/rvck-pipeline/job/check-patch/58/consoleFull

oervci commented:
开始测试

oervci commented:
check patch done. log: https://jenkins.oerv.ac.cn/job/rvck-pipeline/job/check-patch/59/consoleFull

oervci commented:
Kernel build success!

oervci commented:
kunit test done. log:https://jenkins.oerv.ac.cn/job/rvck-pipeline/job/kunit-test/62/consoleFull

oervci commented:
Lava check done! result url: https://lava.oerv.ac.cn/scheduler/job/223

sterling-teng commented:
该补丁backport自kernel社区，代码一致，我赞同合并。
